lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs

```diff
@@ -4131,6 +4207,74 @@ prop_balanceTransactionExistingReturnCollateral
             Left err -> ErrBalanceTxExistingReturnCollateral === err
             e -> counterexample (show e) False
 
+prop_bootstrapWitnesses
+    :: (Word8 -> WriteTx.InAnyRecentEra Cardano.Tx -> Property)
+    -> Word8
+    -- ^ Number of bootstrap witnesses.
+    --
+    -- Testing with [0, 255] should be sufficient.
+    -> AnyRecentEra
+    -> Cardano.NetworkId
+    -- ^ Network - will be encoded inside the witness.
+    -> Index 'WholeDomain 'AccountK
+    -- ^ Account index - will be encoded inside the witness.
+    -> Index 'WholeDomain 'CredFromKeyK
+    -- ^ Index for the first of the 'n' addresses.
+    -> Property
+prop_bootstrapWitnesses
+    p n (AnyRecentEra (era :: RecentEra era)) net accIx addr0Ix =
+    let
+        -- Start incrementing the ixs upward, and if we reach 'maxBound', loop
+        -- around, to ensure we always have 'n' unique indices.
+        addrIxs = take (fromIntegral n)
+            $ [addr0Ix .. maxBound] ++ filter (< addr0Ix) [minBound .. addr0Ix]
+        body = emptyCardanoTxBody
+        wits :: [Cardano.KeyWitness era]
+        wits = map (dummyWitForIx body) addrIxs
+    in
+        p n (WriteTx.InAnyRecentEra era $ Cardano.Tx body wits)
+  where
+    emptyCardanoTxBody = body
+      where
+        Cardano.Tx body _ = WriteTx.toCardanoTx @era $ WriteTx.emptyTx era
+
+    rootK = Byron.generateKeyFromSeed dummyMnemonic mempty
+    pwd = mempty
+
+    dummyWitForIx
+        :: Cardano.TxBody era
+        -> Index 'WholeDomain 'CredFromKeyK
+        -> Cardano.KeyWitness era
+    dummyWitForIx body ix =
+        let
+            accK = Byron.deriveAccountPrivateKey pwd rootK accIx
+            addrKeyAtIx i = Byron.deriveAddressPrivateKey pwd accK i
+
+            addrK = addrKeyAtIx $ toEnum $ fromEnum ix
+            addr = case net of
+                Cardano.Mainnet ->
+                    paymentAddress @'Mainnet $ publicKey addrK
+                Cardano.Testnet _magic ->
+                    -- The choice of network magic here is not important. The
+                    -- size of the witness will not be affected by it. What may
+                    -- affect the size, is the 'Cardano.NetworkId' we pass to
+                    -- 'mkByronWitness' above.
+                    paymentAddress @('Testnet 0) $ publicKey addrK
+        in
+            case era of
+                RecentEraConway ->
+                    mkByronWitness body net addr (getRawKey addrK, pwd)
+                RecentEraBabbage ->
+                    mkByronWitness body net addr (getRawKey addrK, pwd)
+                RecentEraAlonzo  ->
+                    mkByronWitness body net addr (getRawKey addrK, pwd)
```

Comment:
I keep racking my brains as to how we can reduce this kind of duplication.

I hope we can come up with a decent solution eventually!

Reply:
Yes 👍 

Honestly I think I slightly prefer the duplication to reasoning about specific ledger-internal constraints in multiple places.

I would like to continue the discussion about adding super-class constraints to `IsRecentEra` / `RecentEra` though... I'm not sure the confusing missing-instance error message we mentioned last time actually is a problem, or at least not sure they make things worse than how they now are 🤔